### ISSUE TITLE
Update-DbaServiceAccount  with cluster

### ISSUE THREAD

**Issue by jarole** [sqlcollaborative/dbatools#7815]
### Verified issue does not already exist?

No

### What error did you receive?

WARNING: [12:30:17][Update-DbaServiceAccount] Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, System.Object)'." 

### Steps to Reproduce

```powershell
$SecurePw_Agent = ConvertTo-SecureString 'xyz123' -AsPlainText -Force
$SecurePw_Engine = ConvertTo-SecureString 'abc123' -AsPlainText -Force
Update-DbaServiceAccount  -ComputerName NODE2 -ServiceName 'SQLAgent$INSTANCENAME'  -Username    'DOMAIN\agent_user'     -SecurePassword     $SecurePw_Agent -NoRestart
Update-DbaServiceAccount  -ComputerName NODE2 -ServiceName 'MSSQL$INSTANCENAME'     -Username    'DOMAIN\service_user'   -SecurePassword     $SecurePw_Engine  -NoRestart
```

### Are you running the latest release?

Yes

### Other details or mentions

Hi, I'd like to document our error getting during password update on clustered instance (Windows Cluster)
We used most recent dbatools-development , but the same result is in dbatools-master. (23-SEP-2021).
Two years ago, it was working as expected ( `-NoRestart` parameter not used).

When used node name (NODE2), the error occured. 
Using virtual clustername (SQLCLUSTER), the command works, but it means we cannot update passwords on the passive node (stopped service).

Beside this, `-NoRestart` parameter **restarted** our **instance** , what is not expected behavior.

command:
```
$SecurePw_Agent = ConvertTo-SecureString 'xyz123' -AsPlainText -Force
$SecurePw_Engine = ConvertTo-SecureString 'abc123' -AsPlainText -Force
Update-DbaServiceAccount  -ComputerName NODE2 -ServiceName 'SQLAgent$INSTANCENAME'  -Username    'DOMAIN\agent_user'     -SecurePassword     $SecurePw_Agent -NoRestart
Update-DbaServiceAccount  -ComputerName NODE2 -ServiceName 'MSSQL$INSTANCENAME'     -Username    'DOMAIN\service_user'   -SecurePassword     $SecurePw_Engine  -NoRestart
```

Error:
`WARNING: [12:30:17][Update-DbaServiceAccount] Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, System.Object)'." `

### What PowerShell host was used when producing this error

Windows PowerShell ISE (powershell_ise.exe)

### PowerShell Host Version

4.0

### SQL Server Edition and Build number

SQL Server 2012 (SP4-GDR) (KB4583465) - 11.0.7507.2 (X64)  Standard Edition (64-bit)
Windows 2012 R2

**Comment by potatoqualitee** [comment]
Thank you for the report @jarole. Any chance you take a look, @nvarscar ? 🙇🏼

**Comment by nvarscar** [comment]
Could you run the command with `-Verbose` and from the same console where you ran the command run the following:

```powershell
$error[0] | Select *
$error[0].Exception | Select *
$error[0].Exception.InnerException | Select *
```

If  `$error[0]` contains a different error, try `$error[1]`.
Please post all the output.

**Comment by jarole** [comment]
**Legend**:
NODE1 - name of the active node 
NODE2 - name of passive node (stopped SQL service)
SQLCLUSTER - virtual name of the clustered instance

Command has been tested in various ways, executed from both of the nodes, with different results.
We are trying to change password on the passive node NODE2 first, using `-NoRestart` parameter to achieve minimal downtime.

Update-DbaServiceAccount **WORKS (partialy)**
When I run the `Update-DbaServiceAccount` from the active node (NODE1) changing password on the passive node (NODE2).
Password is changed, but the whole clustered instance is stopped and has to be started manually. Even if `-NoRestart` was used. 


Update-DbaServiceAccount **DOESN'T WORK** (error thrown)
When I run the command from the passive node (NODE2), changing password on the passive node (NODE2).
Then we got the error

**VERBOSE OUTPUT**
```
$SecurePw_Agent = ConvertTo-SecureString 'abc123' -AsPlainText -Force
Update-DbaServiceAccount  -ComputerName	NODE2	-ServiceName	'SQLAgent$INSTANCE'	-Username	'DOMAIN\user'	-SecurePassword	$SecurePw_Agent -NoRestart -Verbose


VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving NODE2 using .NET.Dns GetHostEntry
VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving ...IPaddress... using .NET.Dns GetHostByAddress
VERBOSE: [09:57:33][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Retrieving Management Information
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Accessing computer using Cim over DCOM
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Accessing computer using Cim over DCOM - Success
VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving NODE2.fqdn using .NET.Dns GetHostEntry
VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving NODE2 using .NET.Dns GetHostEntry
VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving ...IPaddress... using .NET.Dns GetHostByAddress
VERBOSE: [09:57:33][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Retrieving Management Information
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Accessing computer using Cim over DCOM
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2] Accessing computer using Cim over DCOM - Success
VERBOSE: [09:57:33][Resolve-DbaNetworkName] Resolving NODE2.fqdn using .NET.Dns GetHostEntry
VERBOSE: [09:57:33][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Retrieving Management Information
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM - Success
VERBOSE: [09:57:33][Get-DbaService] Getting SQL Reporting Server services on NODE2.fqdn
VERBOSE: [09:57:33][Get-DbaReportingService] Getting Reporting Server namespace on NODE2.fqdn
VERBOSE: [09:57:33][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Retrieving Management Information
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM
VERBOSE: [09:57:33][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM - Failed
VERBOSE: [09:57:34][Get-DbaReportingService] No SQLServer\ReportServer Namespace on NODE2.fqdn. Please note that this function is available from SQL 2005 up.
VERBOSE: [09:57:34][Get-DbaService] Getting SQL Server namespaces on NODE2.fqdn
VERBOSE: [09:57:34][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:34][Get-DbaCmObject] [NODE2.fqdn] Retrieving Management Information
VERBOSE: [09:57:34][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM
VERBOSE: [09:57:34][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM - Success
VERBOSE: [09:57:34][Get-DbaService] The following namespaces have been found: ComputerManagement11.
VERBOSE: [09:57:34][Get-DbaService] Getting Cim class SqlService in Namespace ComputerManagement11 on NODE2.fqdn.
VERBOSE: [09:57:34][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [09:57:34][Get-DbaCmObject] [NODE2.fqdn] Retrieving Management Information
VERBOSE: [09:57:34][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM
VERBOSE: [09:57:38][Get-DbaCmObject] [NODE2.fqdn] Accessing computer using Cim over DCOM - Success
VERBOSE: [09:57:38][Get-DbaService] Found service SQLAgent$INSTANCE in namespace ComputerManagement11.
VERBOSE: Performing the operation "Changing account information for service SQLAgent$INSTANCE on NODE2.fqdn" on target "SqlService (ServiceName = "SQLAgent?INSTANCE", SQLServiceType = 2)".
VERBOSE: [09:57:38][Update-DbaServiceAccount] Attempting an account change for service SQLAgent$INSTANCE on NODE2.fqdn
VERBOSE: [09:57:38][Resolve-DbaNetworkName] Resolving NODE2 using .NET.Dns GetHostEntry
VERBOSE: [09:57:38][Resolve-DbaNetworkName] Resolving ...IPaddress... using .NET.Dns GetHostByAddress
VERBOSE: [09:57:38][Invoke-ManagedComputerCommand] Connecting to SQL WMI on NODE2.fqdn.
VERBOSE: [09:57:38][Invoke-ManagedComputerCommand] Local connection attempt to NODE2.fqdn failed. Connecting remotely.
WARNING: [09:57:38][Update-DbaServiceAccount] Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, System.Object)'."

------------------------------------------------------
PS C:\Windows\system32> $error[0] | Select *
$error[0].Exception | Select *
$error[0].Exception.InnerException | Select *


writeErrorStream      : True
PSMessageDetails      : 
Exception             : System.Exception: Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, 
                        System.Object, System.Object)'."
TargetObject          : 
CategoryInfo          : NotSpecified: (:) [Write-Error], Exception
FullyQualifiedErrorId : dbatools_Update-DbaServiceAccount,Stop-Function
ErrorDetails          : Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, 
                        System.Object)'."
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Stop-Function, C:\Program Files\WindowsPowerShell\Modules\dbatools\internal\functions\flowcontrol\Stop-Function.ps1: line 260
                        at Update-DbaServiceAccount<End>, C:\Program Files\WindowsPowerShell\Modules\dbatools\functions\Update-DbaServiceAccount.ps1: line 236
                        at <ScriptBlock>, <No file>: line 5
PipelineIterationInfo : {0, 1}
Message        : Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, 
                 System.Object)'."
Data           : {}
InnerException : 
TargetSite     : 
StackTrace     : 
HelpLink       : 
Source         : 
HResult        : -2146233088
```

**Comment by wsmelton** [comment]
Just to cover the bases would you mind testing this from `PowerShell.exe` process as well?

**Comment by jarole** [comment]
> Just to cover the bases would you mind testing this from `PowerShell.exe` process as well?

The result is the same from Powershell_ISE and also from PowerShell.exe

**Comment by nvarscar** [comment]
Let's try to narrow this down to WMI/SMO. Replace variables as necessary:
```powershell
Import-Module dbatools
$serviceName = 'SQLAgent$INSTANCE'
$username = 'My\UserName'
$password = 'myPassword'
$fqdn = 'NODE2.domain.name'
[void][System.Reflection.Assembly]::LoadWithPartialName('Microsoft.SqlServer.SqlWmiManagement')
$wmi = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $fqdn
$null = $wmi.Initialize()
$service = $wmi.Services[$serviceName]
$service.SetServiceAccount($username, $password)
$service.Alter()
```

The above should theoretically throw the same error. I think, it needs to be executed in an elevated console, but that might be the localhost-only requirement. Just in case, try both.

**Comment by jarole** [comment]
Your'e right, it throws the same error. 
We always run the commands in elevated console (but for this case I tested both - no difference)

```
New-Object : Exception calling ".ctor" with "1" argument(s): "Method not found: 'System.String System.String.Format(System.IFormatProvider, System.String, System.Object, System.Object)'."
At line:7 char:8
+ $wmi = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $fqdn
+        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [New-Object], MethodInvocationException
    + FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCommand
 
You cannot call a method on a null-valued expression.
At line:8 char:1
+ $null = $wmi.Initialize()
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Cannot index into a null array.
At line:9 char:1
+ $service = $wmi.Services[$serviceName]
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray
 
You cannot call a method on a null-valued expression.
At line:10 char:1
+ $service.SetServiceAccount($username, $password)
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:1
+ $service.Alter()
+ ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
```

**Comment by wsmelton** [comment]
@jarole are you running this from a node on the cluster?

If so what version of .NET Framework do you have installed?

**Comment by jarole** [comment]
@wsmelton , yes I run this from a cluster node. The result depends on which node I run the command from. As described in my second post https://github.com/sqlcollaborative/dbatools/issues/7815#issuecomment-929975030

tldr;
- when running from **passive** node (changing password on the **passive** node) = the error occures
- when running from **active** node (changing password on the **passive** node) = password changed, but restart occures 
( `-NoRestart` parameter used)

.NET version installed is 4.5

**Comment by wsmelton** [comment]
> .NET version installed is 4.5

As of version 1.1.0 release ([ref](https://github.com/sqlcollaborative/dbatools/releases/tag/v1.1.0)) the module requires .NET 4.6.2 to be installed on your server. Would you be able to install that on one of the nodes and retest?

**Comment by jarole** [comment]
Thank you.
Next days, it is not possible, but once installed I will test and inform you.
